### PR TITLE
Add support for password protected archives

### PR DESCRIPTION
**Pull request by Vadiml1024**
This pull request combines patches from https://github.com/richnetdesign/python-libarchive
And my work to implement handling of password protected archives and building against latest version of libachive

### CODE REVIEW

**Comment by cabarnes** [comment]
Can you update this with `master`? I moved the checks to GitHub actions so they will actually run

**Comment by Vadiml1024** [comment]
Done

**Comment by Vadiml1024** [comment]
Do i need re-submit the pull req?

**Comment by cabarnes** [comment]
It looks like there's an error when running the tests in Python 2

**Comment by Vadiml1024** [comment]
I've fixed it

File: .github/workflows/build-wheels.yaml [hunk 1]
```diff
@@ -0,0 +1,104 @@
+name: build-wheels
```
Comment (cabarnes) [inline .github/workflows/build-wheels.yaml]: This file seems to have come from the forked repo. We'll put in our own workflow for deploying to PyPI. Can this file be removed?

File: Makefile [hunk 2]
```diff
@@ -11,6 +11,9 @@ verify:
 install:
 	python setup.py install
 
+wheel:
```
Comment (cabarnes) [inline Makefile]: This target can be removed as well

File: libarchive/Makefile [hunk 3]
```diff
@@ -1,19 +1,24 @@
-CFLAGS     = -g
-INCLUDE    = -I/usr/include -I.
-LIBS       = -larchive
+CFLAGS     = -g
```
Comment (cabarnes) [inline libarchive/Makefile]: Trailing space should be removed

File: libarchive/Makefile [hunk 4]
```diff
@@ -1,19 +1,24 @@
...
 
-PYVER ?= 2.7
+#DEFINES = -DPyUnicode_AS_STRING=PyUnicode_AS_DATA
+PYVER ?= 3.9
 
 all: __libarchive.so
 
 _libarchive_wrap.c: _libarchive.i
-	swig -python -shadow _libarchive.i
+	#swig -python -c++ -Wall -shadow _libarchive.i
```
Comment (cabarnes) [inline libarchive/Makefile]: Can the commented out lines in this file be removed?

File: libarchive/Makefile [hunk 5]
```diff
@@ -1,19 +1,24 @@
...
 
 _libarchive_wrap.o: _libarchive_wrap.c
-	${CC} -c ${CFLAGS} -fPIC ${INCLUDE} $$(python${PYVER}-config --cflags) _libarchive_wrap.c
+	echo ${CC} -c ${CFLAGS} -fPIC  $$(python${PYVER}-config --cflags) _libarchive_wrap.c
+	${CC} -c ${CFLAGS} -fPIC  $$(python${PYVER}-config --cflags) ${DEFINES} _libarchive_wrap.c
+#	${CXX} -c ${CFLAGS} -fPIC  $$(python${PYVER}-config --cflags) -DPyUnicode_AS_STRING=PyUnicode_AS_DATA _libarchive_wrap.cxx
 
 __libarchive.so: _libarchive_wrap.o
-	${CC} _libarchive_wrap.o -shared $$(python${PYVER}-config --ldflags) -o __libarchive.so ${LIBS}
+	echo INCLUDE=$INCLUDE LIBS=$LIBS
```
Comment (cabarnes) [inline libarchive/Makefile]: This looks like it was for debugging and can removed

File: libarchive/__init__.py [hunk 6]
```diff
@@ -29,10 +29,12 @@
 import time
 import warnings
 
+# from ctypes import cdll, c_char_p
```
Comment (cabarnes) [inline libarchive/__init__.py]: Can this be removed?

File: libarchive/__init__.py [hunk 7]
```diff
@@ -295,14 +310,24 @@ def close(self):
...
         self.pathname = pathname
         self.size = size
         self.mtime = mtime
         self.mode = mode
         self.hpos = hpos
         self.encoding = encoding
 
+        self.symlink = ""
+
+        # if self.issym() and symlink:
```
Comment (cabarnes) [inline libarchive/__init__.py]: Can this be removed?

File: libarchive/__init__.py [hunk 8]
```diff
@@ -356,6 +386,10 @@ def from_file(cls, f, entry=None, encoding=ENCODING):
                 entry.size = getattr(f, 'size', 0)
                 entry.mtime = getattr(f, 'mtime', time.time())
                 entry.mode = stat.S_IFREG
+
+            if stat.S_ISLNK(entry.mode):
+                print("yo")
```
Comment (cabarnes) [inline libarchive/__init__.py]: This looks like debugging code that can be removed

File: libarchive/__init__.py [hunk 9]
```diff
@@ -370,8 +404,14 @@ def to_archive(self, archive):
...
             _libarchive.archive_entry_set_size(e, self.size)
             _libarchive.archive_entry_set_mtime(e, self.mtime, 0)
+
+            if stat.S_ISLNK(self.mode):
+                _libarchive.archive_entry_set_symlink(e, self.symlink)
+
             call_and_check(_libarchive.archive_write_header, archive._a, archive._a, e)
-            #self.hpos = archive.header_position
+
+            # todo
```
Comment (cabarnes) [inline libarchive/__init__.py]: Does this need to be done?

File: setup.py [hunk 10]
```diff
@@ -34,30 +34,45 @@
     from distutils.core import setup, Extension
     from distutils.command.build_ext import build_ext
 
+import subprocess
```
Comment (cabarnes) [inline setup.py]: Does this file need to be changed? It looks like some of it is particular to what you were doing (`repourl = 'https://github.com/Vadiml1024/python-libarchive'`)

File: pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+# See https://setuptools.readthedocs.io/en/latest/build_meta.html
```
Comment (cabarnes) [inline pyproject.toml]: We didn't use this before and aren't planning on using that build tool. Can this be removed?

File: tests.py [hunk 12]
```diff
@@ -33,7 +33,8 @@
 from libarchive import Archive, is_archive_name, is_archive
 from libarchive.zip import is_zipfile, ZipFile, ZipEntry
 
-PY3 = sys.version_info[0] == 3
+# PY3 = sys.version_info[0] == 3
+PY3 = True
```
Comment (cabarnes) [inline tests.py]: We run the tests in python 2 and 3. Why shouldn't this detect that?

File: tests.py [hunk 13]
```diff
@@ -175,8 +176,15 @@ def test_writepath(self):
                 z.writepath(f)
         z.close()
 
+    def test_writepath_with_password(self):
+        z = ZipFile(self.f, 'w', password='test')
+        for fname in FILENAMES:
+            with open(os.path.join(TMPDIR, fname), 'r') as f:
+                z.writepath(f)
+        z.close()
```
Comment (cabarnes) [inline tests.py]: Is there a way to verify that the created file requires a password to open now?

**Review by cabarnes — CHANGES_REQUESTED** [review]

File: .github/workflows/build-wheels.yaml (hunk 1 above)
Comment (Vadiml1024) [reply .github/workflows/build-wheels.yaml]: Done

File: Makefile (hunk 2 above)
Comment (Vadiml1024) [reply Makefile]: Done

File: libarchive/Makefile (hunk 4 above)
Comment (Vadiml1024) [reply libarchive/Makefile]: Done

File: libarchive/__init__.py (hunk 9 above)
Comment (Vadiml1024) [reply libarchive/__init__.py]: Done

File: pyproject.toml (hunk 11 above)
Comment (Vadiml1024) [reply pyproject.toml]: I copied it from som template do you want me to remove those file altogether?

File: tests.py (hunk 12 above)
Comment (Vadiml1024) [reply tests.py]: Done

File: pyproject.toml (hunk 11 above)
Comment (cabarnes) [reply pyproject.toml]: Yes, just remove it

Comment (Vadiml1024) [reply pyproject.toml]: Done

File: setup.py (hunk 10 above)
Comment (Vadiml1024) [reply setup.py]: Done

File: libarchive/__init__.py (hunk 6 above)
Comment (Vadiml1024) [reply libarchive/__init__.py]: Done

File: pyproject.toml (hunk 11 above)
Comment (Vadiml1024) [reply pyproject.toml]: OK

**Comment by Vadiml1024** [comment]
I Think i've fixed all issues

File: libarchive/_libarchive.i [hunk 14]
```diff
@@ -171,39 +169,107 @@ extern int archive_read_next_header2(struct archive *,
...
+ * Add a decryption passphrase.
+ */
+extern int archive_read_add_passphrase(struct archive *, const char *);
 /* data */
 extern int archive_read_data_skip(struct archive *);
 extern int archive_read_data_into_fd(struct archive *, int fd);
 
-/* FILTERS */
+#if ARCHIVE_VERSION_NUMBER < 4000000
+extern int archive_read_support_compression_all(struct archive *);
```
Comment (cabarnes) [inline libarchive/_libarchive.i:200]: When I build, I get warnings that all of these are deprecated
```
cc -c -g -fPIC  $(python3.8-config --cflags)  _libarchive_wrap.c
_libarchive_wrap.c: In function ‘_wrap_archive_read_support_compression_all’:
_libarchive_wrap.c:3864:3: warning: ‘archive_read_support_compression_all’ is deprecated [-Wdeprecated-declarations]
 3864 |   result = (int)archive_read_support_compression_all(arg1);
      |   ^~~~~~
In file included from _libarchive_wrap.c:2712:
/usr/include/archive.h:392:15: note: declared here
  392 | __LA_DECL int archive_read_support_compression_all(struct archive *)
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```
Are you seeing the same thing? Is there an updated interface that should be used? It looks like there is a check for `ARCHIVE_VERSION_NUMBER` that should be accounting for that

Comment (Vadiml1024) [reply libarchive/_libarchive.i:200]: Yes i'm getting them too.
Given the fact that the master branch version archive.h in libarchive repo is only   3.6.2
I think we can keep it as it is...

File: tests.py [hunk 15]
```diff
@@ -246,12 +247,64 @@ def test_deferred_close_by_archive(self):
...
+ZIP_CONTENT='UEsDBAoACQAAAKubCVVjZ7b1FQAAAAkAAAAIABwAdGVzdC50eHRVVAkAA5K18mKStfJid' + \
+        'XgLAAEEAAAAAAQAAAAA5ryoP1rrRK5apjO41YMAPjpkWdU3UEsHCGNntvUVAAAACQAAAF' + \
+        'BLAQIeAwoACQAAAKubCVVjZ7b1FQAAAAkAAAAIABgAAAAAAAEAAACkgQAAAAB0ZXN0LnR' + \
+        '4dFVUBQADkrXyYnV4CwABBAAAAAAEAAAAAFBLBQYAAAAAAQABAE4AAABnAAAAAAA='
+
+ITEM_CONTENT='test.txt\n'
+ITEM_NAME='test.txt'
+
+
+class TestPasswordProtection(unittest.TestCase):
```
Comment (cabarnes) [inline tests.py]: Can you add a test for creating an archive with a password?

Comment (Vadiml1024) [reply tests.py]: Done

File: libarchive/__init__.py [hunk 16]
```diff
@@ -28,85 +28,124 @@
...
 from io import StringIO
 
-PY3 = sys.version_info[0] == 3
+PY3 = sys.version_info[0] >= 3
 
 # Suggested block size for libarchive. Libarchive may adjust it.
 BLOCK_SIZE = 10240
 
-MTIME_FORMAT = ''
+MTIME_FORMAT = ""
```
Comment (cabarnes) [inline libarchive/__init__.py]: It seems like this whole file was just run through some kind of auto formatter. Can you limit the changes to only what is relevant to this change?

**Review by cabarnes — CHANGES_REQUESTED** [review]

File: libarchive/__init__.py (hunk 16 above)
Comment (Vadiml1024) [reply libarchive/__init__.py]: Sorry about that...
I'm newbie with VSCode...  It proposed to install some python linting plugins and I've accepted.
I don't really know how to go back and undo formatting it did....

Comment (Vadiml1024) [reply libarchive/__init__.py]: I've fixed the formatinng issue

File: libarchive/Makefile [hunk 17]
```diff
@@ -1,19 +1,20 @@
 CFLAGS     = -g
-INCLUDE    = -I/usr/include -I.
-LIBS       = -larchive
+INCLUDE    = -I/usr/local/include -I/usr/include -I.
+LIBS       = -L /usr/local/lib -larchive
 
-PYVER ?= 2.7
+#DEFINES = -DPyUnicode_AS_STRING=PyUnicode_AS_DATA
```
Comment (cabarnes) [inline libarchive/Makefile]: Can this be removed and the place below that it was used?

File: libarchive/__init__.py [hunk 18]
```diff
@@ -295,14 +304,19 @@ def close(self):
 
 class Entry(object):
     '''An entry within an archive. Represents the header data and it's location within the archive.'''
+
     def __init__(self, pathname=None, size=None, mtime=None, mode=None, hpos=None, encoding=ENCODING):
+
+        # , symlink=None
```
Comment (cabarnes) [inline libarchive/__init__.py]: Can this be removed?

File: setup.py [hunk 19]
```diff
@@ -35,29 +35,38 @@
...
+else:
+    includePath = '/usr/local/include'
+
+
+
+
 name = 'python-libarchive'
 version = '4.0.1'
-release = '1'
+release = '2'
```
Comment (cabarnes) [inline setup.py]: Can this be set back to 1? I'll change the version when doing the release to PyPI

**Review by cabarnes — COMMENTED** [review]
Everything looks good. Just a couple more little cleanups and then this should be good to merge

File: setup.py (hunk 19 above)
Comment (Vadiml1024) [reply setup.py]: Done

File: libarchive/Makefile (hunk 17 above)
Comment (Vadiml1024) [reply libarchive/Makefile]: Done

File: libarchive/__init__.py (hunk 18 above)
Comment (Vadiml1024) [reply libarchive/__init__.py]: Done

**Review by cabarnes — APPROVED** [review]
Thanks for all the hard work to add this feature!

**Comment by Vadiml1024** [comment]
I'm having problem to use python-libarchive=4.1.0  on systems with libarchive.so.13 because
this version of libarchive does not have archive_write_set_passphrase function
This is the reason when initially my github workflow built libacrchive.so itself and embedded it and dependencies in the wheel.

**Comment by cabarnes** [comment]
I am able to use it to create password protected archives on Ubuntu 20.04 and 22.04 without issue. They are using `libarchive.so.13.4.0` and `libarchive.so.13.6.0` respectively

**Pull request merged**